core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -858,13 +858,22 @@ class Partition(val topicPartition: TopicPartition,
     // No need to calculate low watermark if there is no delayed DeleteRecordsRequest
     val oldLeaderLW = if (delayedOperations.numDelayedDelete > 0) lowWatermarkIfLeader else -1L
     val prevFollowerEndOffset = replica.stateSnapshot.logEndOffset
-    replica.updateFetchState(
-      followerFetchOffsetMetadata,
-      followerStartOffset,
-      followerFetchTimeMs,
-      leaderEndOffset,
-      brokerEpoch
-    )
+
+    // Acquire the lock for the fetch state update. A race can happen between fetch requests from a rebooted broker.
+    // The requests before and after the reboot can carry different fetch metadata especially offsets and broker epoch.
+    // It can particularly affect the ISR expansion where we decide to expand based on stale fetch request but use the
+    // latest broker epoch to fill in the AlterPartition request.
+    inReadLock(leaderIsrUpdateLock) {
+      if (!replica.maybeUpdateFetchState(
+        followerFetchOffsetMetadata,
+        followerStartOffset,
+        followerFetchTimeMs,
+        leaderEndOffset,
+        brokerEpoch
+      )) {
+        return
```

Comment:
Should we return an error here? `followerReplicaOrThrow` may be a good inspiration for the errors that we could use here.